File: apps/zxtune-android/zxtune/src/test/java/app/zxtune/core/IdentifierTest.java
package app.zxtune.core;

import static org.junit.Assert.assertEquals;

import android.net.Uri;

import org.junit.Test;
import org.junit.runner.RunWith;
import org.robolectric.RobolectricTestRunner;

@RunWith(RobolectricTestRunner.class)
public class IdentifierTest {
  @Test
  public void parseEmpty() {
    Identifier id = Identifier.parse(null);

    assertEquals("", id.toString());
    assertEquals(Uri.EMPTY, id.getFullLocation());
    assertEquals(Uri.EMPTY, id.getDataLocation());
    assertEquals("", id.getSubPath());
    assertEquals("", id.getDisplayFilename());

    verifyEquals(id, Identifier.parse(""));
    verifyEquals(id, new Identifier(Uri.EMPTY));
    verifyEquals(id, new Identifier(Uri.EMPTY, null));
    verifyEquals(id, new Identifier(Uri.EMPTY, ""));
    verifyEquals(id, new Identifier(Uri.EMPTY, "ignored"));
  }

  private static void verifyEquals(Identifier lh, Identifier rh) {
    assertEquals(lh, rh);
    assertEquals(lh.toString(), rh.toString());
  }

  @Test
  public void parseFile() {
    String path = "dir/file";
    Identifier id = Identifier.parse(path);

    assertEquals(path, id.toString());
    assertEquals(path, id.getFullLocation().toString());
    assertEquals(path, id.getDataLocation().toString());
    assertEquals("", id.getSubPath());
    assertEquals("file", id.getDisplayFilename());

    verifyEquals(id, Identifier.parse(path + "#"));
    verifyEquals(id, new Identifier(Uri.parse(path)));
    verifyEquals(id, new Identifier(Uri.parse(path), null));
    verifyEquals(id, new Identifier(Uri.parse(path), ""));
  }

  @Test
  public void parseFull() {
    String path = "http://host/path?query";
    String subPath = "nested/sub path";
    String subPathEncoded = Uri.encode(subPath);

    Identifier id = Identifier.parse(path + '#' + subPathEncoded);

    assertEquals(path + '#' + subPathEncoded, id.toString());
    assertEquals(path + '#' + subPathEncoded, id.getFullLocation().toString());
    assertEquals(path, id.getDataLocation().toString());
    assertEquals("nested/sub path", id.getSubPath());
    assertEquals("path > sub path", id.getDisplayFilename());

    verifyEquals(id, new Identifier(Uri.parse(path), subPath));
  }
}

File: apps/zxtune-android/zxtune/src/test/java/app/zxtune/core/IdentifierTest.kt
package app.zxtune.core

import android.net.Uri
import org.junit.Assert.assertEquals
import org.junit.Test
import org.junit.runner.RunWith
import org.robolectric.RobolectricTestRunner

private fun verifyEquals(expected: Identifier, actual: Identifier) {
    assertEquals(expected, actual)
    assertEquals(expected.toString(), actual.toString())
}

@RunWith(RobolectricTestRunner::class)
class IdentifierTest {
    @Test
    fun `test parsing empty`() = with(Identifier.parse(null)) {

        assertEquals("", toString())
        assertEquals(Uri.EMPTY, fullLocation)
        assertEquals(Uri.EMPTY, dataLocation)
        assertEquals("", subPath)
        assertEquals("", displayFilename)
        assertEquals(null, trackIndex)

        verifyEquals(this, Identifier.parse(""))
        verifyEquals(this, Identifier(Uri.EMPTY))
        verifyEquals(this, Identifier(Uri.EMPTY, null))
        verifyEquals(this, Identifier(Uri.EMPTY, ""))
        verifyEquals(this, Identifier(Uri.EMPTY, "ignored"))
    }

    @Test
    fun `test parsing file path`() {
        val path = "dir/file"
        with(Identifier.parse(path)) {

            assertEquals(path, toString())
            assertEquals(path, fullLocation.toString())
            assertEquals(path, dataLocation.toString())
            assertEquals("", subPath)
            assertEquals("file", displayFilename)
            assertEquals(null, trackIndex)

            verifyEquals(this, Identifier.parse("${path}#"))
            verifyEquals(this, Identifier(Uri.parse(path)))
            verifyEquals(this, Identifier(Uri.parse(path), null))
            verifyEquals(this, Identifier(Uri.parse(path), ""))
        }
    }

    @Test
    fun `test parsing full uri`() {
        val path = "http://host/path?query"
        val subPath = "nested/sub path"
        val subPathEncoded = Uri.encode(subPath)

        with(Identifier.parse("${path}#${subPathEncoded}")) {

            assertEquals("${path}#${subPathEncoded}", toString())
            assertEquals("${path}#${subPathEncoded}", fullLocation.toString())
            assertEquals(path, dataLocation.toString())
            assertEquals("nested/sub path", this.subPath)
            assertEquals("path > sub path", displayFilename)
            assertEquals(null, trackIndex)

            verifyEquals(this, Identifier(Uri.parse(path), subPath))
        }
    }

    @Test
    fun `test parsing full url with track index`() {
        val path = "http://host/path?query"
        val subPath = "nested/sub path/#123"
        val subPathEncoded = Uri.encode(subPath)

        with(Identifier.parse("${path}#${subPathEncoded}")) {

            assertEquals("${path}#${subPathEncoded}", toString())
            assertEquals("${path}#${subPathEncoded}", fullLocation.toString())
            assertEquals(path, dataLocation.toString())
            assertEquals("nested/sub path/#123", this.subPath)
            assertEquals("path > #123", displayFilename)
            assertEquals(123, trackIndex)

            verifyEquals(this, Identifier(Uri.parse(path), subPath))
        }
    }
}
